### DISCUSSION TITLE
Integration with Rust 3D Graphics Libraries

### DISCUSSION THREAD

**Discussion by JohnPiwinski** [slint-ui/slint#5599] [Q&A]
I am attempting to write a tool to assist doctors with visualizing tissue biopsies that have been stained with multiple fluorescent dyes to identify cell types. How difficult would it be to include a 3D graphics window in a Slint application for this purpose? I have some experience using [this library](https://github.com/asny/three-d), and I heard someone mention that I could employ glium, wgpu, or winit directly to write the 3D content to the graphics window. I am afraid that if I do not use an official API, the code will be fragile and make future updates prohibitively difficult.

Thank you very much

**Comment by tronical** [comment]
Hi! I Think this should be quite doable, three-d looks very promising and flexible. Depending on how you'd like to integrate the rendering into your UI, I think the https://github.com/slint-ui/slint/tree/master/examples/opengl_texture and the https://github.com/slint-ui/slint/tree/master/examples/opengl_underlay might serve as starting points. Slint provides a current OpenGL context in the rendering notifier callbacks, along with a get_proc_address function pointer to initialise your own GL wrapper.

With three-d, I don't think you can use the `window` module, but [`Context`](https://docs.rs/three-d/0.17.0/three_d/context/struct.Context.html#method.from_loader_function_cstr) has the right signature that matches what we provide.

**Reply by JohnPiwinski** [reply]
Thanks so much! I'll look into these examples and provide an update if I get a working demo.

**Reply by GayarUsmanov** [reply]
> Thanks so much! I'll look into these examples and provide an update if I get a working demo.

Hi @JohnPiwinski,
How did it go?

**Comment by vigilanz** [comment]
I just figured this out and posted the answer to my own question in the disussions over at three-d. the link is https://github.com/asny/three-d/discussions/509
So far, nobody there has commented on it (though I just posted it) - but it works so it cannot be thaat wrong.
